File: php/removeElement.php
<?php

class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $val
     * @return Integer
     */
    function removeElement(&$nums, $val) {
        $i = 0;
        $j = count($nums) -1;

        if($i === $j) {
            if($nums[$i] == $val) {
                $nums = [];
            }
        }else {
            while($i < $j) {
                $curr = $nums[$i];
                $last = $nums[$j];

                if($curr === $val) {
                    if ($last === $val) {
                        while ($j > $i) {
                            if($nums[$j] !== $val) {
                                break;
                            }
                            // $nums[$j] = null;
                            $j--;
                        }
                    }else {
                        // $tmp = $curr;
                        $nums[$i] = $last;
                        $nums[$j] = null;
                        $i++;
                        $j--;
                    }
                }else {
                    $i++;
                }
            }
            $nums = array_slice($nums, 0, $j+1);
        }
    }

    function newRemoveElement(&$nums, $val) {
        $i = 0;
        for ($j = 0; $j < count($nums); $j++) {
            if ($nums[$j] != $val) {
                $nums[$i] = $nums[$j];
                $i++;
            }
        }
        return $i;
    }
}


$nums = [0,1,2,2,3,0,4,2];
$val = 2;

$solution = new Solution();
$solution->newRemoveElement($nums, $val);

print_r($nums);

File: 2025/leetcode /removeElement.py
from typing import List

# class Solution:
#     def removeElement(self, nums: List[int], val: int) -> int:
#       i = 0
#       j = len(nums) - 1
#       m = 0

#       for num in nums:
#         if num == val:
#           m += 1

#       while i < j:
#         first = nums[i]
#         last = nums[j]

#         if first != val:
#           i += 1
        
#         if last == val:
#           j -= 1

#         if first == val and last != val:
#           nums[i], nums[j] = nums[j], nums[i]
#           i += 1
#           j -= 1
#       return len(nums) - m

class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
      i = 0
      j = len(nums) - 1

      while i <= j:
        first = nums[i]
        last = nums[j]

        if first != val:
          i += 1
        else:
          nums[i] = last
          j -= 1
        
      return i

# class Solution:
#     def removeElement(self, nums: List[int], val: int) -> int:
#       m = 0

#       for i in range(len(nums)):
#         if nums[i] != val:
#           nums[m] = nums[i]
#           m += 1

#       return m 



nums = [0,1,2,2,3,0,4,2]

val = 2

s = Solution()
print(s.removeElement(nums, val))